File: dart/sdk/lib/src/domains.dart
import 'dart:async';

import 'service_client.dart';

final class AuthClient {
  AuthClient(
    this.service, {
    this.realm = '2finance',
    this.clientId = '2finance-network',
    this.phoneClientId = '2finance-network-phone',
  });

  final ServiceClient service;
  final String realm;
  final String clientId;
  final String phoneClientId;

  Future<Object?> login(Map<String, Object?> input) {
    return service.post(_authPath(clientId, '/login'), input);
  }

  Future<Object?> refreshToken(String refreshToken) {
    return service.post(_authPath(clientId, '/refresh'), {
      'refresh_token': refreshToken,
    });
  }

  Future<Object?> phoneLogin(String phoneNumber, String code) {
    return service.post(_authPath(phoneClientId, '/phone/sms/login'), {
      'phone_number': phoneNumber,
      'code': code,
    });
  }

  Future<Object?> jwks() {
    return service.get(_oidcPath('/protocol/openid-connect/certs'));
  }

  Future<Object?> validateToken(String token) {
    return service.post(
      _oidcPath('/protocol/openid-connect/token/introspect'),
      {'token': token},
    );
  }

  String _authPath(String selectedClientId, String endpoint) {
    return '/v1/2finance-authenticator/$realm/$selectedClientId/${endpoint.replaceFirst(RegExp(r'^/+'), '')}';
  }

  String _oidcPath(String endpoint) {
    return '/realms/$realm/${endpoint.replaceFirst(RegExp(r'^/+'), '')}';
  }
}

final class AnalyticsClient {
  AnalyticsClient(this.service);

  final ServiceClient service;

  Future<Object?> indicators() => service.get('/analytics/indicators');

  Future<Object?> calculateTechnicalAnalysis(Map<String, Object?> request) {
    return service.post('/analytics/technical-analysis:calculate', request);
  }

  Future<Object?> optimizePortfolio(Map<String, Object?> request) {
    return service.post('/portfolio-manager/optimizer', request);
  }

  Future<Object?> upsertCandles(Map<String, Object?> request) {
    return service.post('/analytics/candles:upsert', request);
  }

  Future<Object?> rankings() => service.get('/portfolio-manager/rankings');

  Future<Object?> balances(String accountId) {
    return service.get(
      '/portfolio-manager/balances/${Uri.encodeComponent(accountId)}',
    );
  }

  Future<Object?> blackScholes([String query = '']) {
    final suffix = query.isEmpty ? '' : '?$query';
    return service.get('/risk-manager/blackscholes$suffix');
  }

  Future<Object?> staking() => service.get('/staking');
}

final class MCPClient {
  MCPClient(this.service);

  final ServiceClient service;
  int _nextId = 1;

  Future<Object?> call(String method, [Object? params]) {
    return service.post('/mcp', {
      'jsonrpc': '2.0',
      'id': _nextId++,
      'method': method,
      'params': params,
    });
  }

  Future<Object?> listTools() => call('tools/list');

  Future<Object?> listPrompts() => call('prompts/list');

  Future<Object?> listResources() => call('resources/list');

  Future<Object?> readResource(String uri) {
    return call('resources/read', {'uri': uri});
  }

  Future<Object?> callTool(String name, [Map<String, Object?>? arguments]) {
    return call('tools/call', {'name': name, 'arguments': arguments ?? {}});
  }

  Future<Object?> getPrompt(String name, [Map<String, Object?>? arguments]) {
    return call('prompts/get', {'name': name, 'arguments': arguments ?? {}});
  }

  Future<Object?> conversationPlan(Map<String, Object?> arguments) {
    return callTool('finance_assistant.conversation.plan', arguments);
  }
}

final class NetworkClient {
  NetworkClient(this.service);

  final ServiceClient service;

  Future<Object?> virtualMachine() {
    return service.post('/v2/2finance-network/query', {
      'method': 'get_blocks',
      'params': {'page': 1, 'limit': 1},
    });
  }

  Future<Object?> marketCandles(String market, [String query = '']) {
    final suffix = query.isEmpty ? '' : '?$query';
    return service.get(
      '/v2/2finance-network/markets/${Uri.encodeComponent(market)}/candles$suffix',
    );
  }

  Future<Object?> products(String productType) {
    return service.get(
      '/v2/2finance-network/products/${Uri.encodeComponent(productType)}',
    );
  }

  Future<Object?> createProduct(
    String productType,
    Map<String, Object?> request,
  ) {
    return service.post(
      '/v2/2finance-network/products/${Uri.encodeComponent(productType)}',
      request,
    );
  }

  Future<Object?> bonds() => products('bonds');

  Future<Object?> createBond(Map<String, Object?> request) {
    return createProduct('bonds', request);
  }

  Future<Object?> loans() => products('loans');

  Future<Object?> createLoan(Map<String, Object?> request) {
    return createProduct('loans', request);
  }

  Future<Object?> swaps() => products('swaps');

  Future<Object?> createSwap(Map<String, Object?> request) {
    return createProduct('swaps', request);
  }

  Future<Object?> stakingProducts() => products('staking');

  Future<Object?> createStakingProduct(Map<String, Object?> request) {
    return createProduct('staking', request);
  }

  Future<Object?> syntheticAssets() => products('synthetic-assets');

  Future<Object?> createSyntheticAsset(Map<String, Object?> request) {
    return createProduct('synthetic-assets', request);
  }

  Future<Object?> liquidityPools() => products('liquidity-pools');

  Future<Object?> createLiquidityPool(Map<String, Object?> request) {
    return createProduct('liquidity-pools', request);
  }
}

final class OrchestratorClient {
  OrchestratorClient(this.service);

  final ServiceClient service;

  Future<Object?> catalog() => service.get('/v1/mcphost/catalog/packages');

  Future<Object?> createSession(Map<String, Object?> request) {
    return service.post('/v1/mcphost/sessions', request);
  }

  Future<Object?> sendMessage(Map<String, Object?> request) {
    return service.post('/v1/mcphost/messages', request);
  }

  Future<Object?> tools() => service.get('/v1/mcphost/tools');

  Future<Object?> prompts() => service.get('/v1/mcphost/prompts');

  Future<Object?> resources() => service.get('/v1/mcphost/resources');

  Future<Object?> providers() => service.get('/v1/mcphost/providers');

  Future<Object?> approvals() => service.get('/v1/mcphost/approvals');

  Future<Object?> observability() => service.get('/v1/mcphost/observability');

  Future<Object?> deleteSession(String id) {
    return service.delete('/v1/mcphost/sessions/${Uri.encodeComponent(id)}');
  }
}

final class TradingControlClient {
  TradingControlClient(this.service);

  final ServiceClient service;

  Future<Object?> robots() => service.get('/robots');

  Future<Object?> createRobot(Map<String, Object?> request) {
    return service.post('/robots', request);
  }

  Future<Object?> robot(String id) {
    return service.get('/robots/${Uri.encodeComponent(id)}');
  }

  Future<Object?> startRobot(String id) {
    return service.post('/robots/${Uri.encodeComponent(id)}:start');
  }

  Future<Object?> pauseRobot(String id) {
    return service.post('/robots/${Uri.encodeComponent(id)}:pause');
  }

  Future<Object?> resumeRobot(String id) {
    return service.post('/robots/${Uri.encodeComponent(id)}:resume');
  }

  Future<Object?> stopRobot(String id) {
    return service.post('/robots/${Uri.encodeComponent(id)}:stop');
  }

  Future<Object?> riskPolicy(String id) {
    return service.get('/robots/${Uri.encodeComponent(id)}/risk-policy');
  }

  Future<Object?> setRiskPolicy(String id, Map<String, Object?> request) {
    return service.put(
      '/robots/${Uri.encodeComponent(id)}/risk-policy',
      request,
    );
  }

  Future<Object?> riskView(String id) {
    return service.get('/risk-view/${Uri.encodeComponent(id)}');
  }

  Future<Object?> strategies() => service.get('/strategies');

  Future<Object?> createStrategy(Map<String, Object?> request) {
    return service.post('/strategies', request);
  }

  Future<Object?> directives() => service.get('/directives');

  Future<Object?> createDirective(Map<String, Object?> request) {
    return service.post('/directives', request);
  }

  Future<Object?> audit() => service.get('/audit');

  Future<Object?> activity() => service.get('/activity');

  Future<Object?> mcpTools() => service.get('/mcp/tools');
}

final class KeyStoreClient {
  KeyStoreClient(this.service);

  final ServiceClient service;

  Future<Object?> health() => service.get('/healthz');

  Future<Object?> readiness() => service.get('/readyz');

  Future<Object?> startKeygen(Map<String, Object?> request) {
    return service.post('/keystore/keygen/start', request);
  }

  Future<Object?> keygenSignature(Map<String, Object?> request) {
    return service.post('/keystore/keygen/signature', request);
  }

  Future<Object?> startSigning(Map<String, Object?> request) {
    return service.post('/keystore/signing/start', request);
  }

  Future<Object?> signingSignature(Map<String, Object?> request) {
    return service.post('/keystore/signing/signature', request);
  }

  Future<Object?> startResharing(Map<String, Object?> request) {
    return service.post('/keystore/resharing/start', request);
  }

  Future<Object?> keys(String userPublicKey) {
    return service.get('/keystore/keys/${Uri.encodeComponent(userPublicKey)}');
  }

  Future<Object?> signatures(String userPublicKey) {
    return service.get(
      '/keystore/signatures/${Uri.encodeComponent(userPublicKey)}',
    );
  }

  Future<Object?> metrics() => service.get('/keystore/tss/metrics');
}

final class HummingbotClient {
  HummingbotClient(this.service);

  final ServiceClient service;

  Future<Object?> assets() => service.get('/api/v1/assets');

  Future<Object?> symbols() => service.get('/api/v1/symbols');

  Future<Object?> balances() => service.get('/api/v1/balances');

  Future<Object?> connectorConfig(Map<String, Object?> request) {
    return service.post('/api/v1/connectors/2finance/config', request);
  }
}

final class MatchEngineClient {
  MatchEngineClient(this.webSocketUrl);

  final String webSocketUrl;

  Map<String, Object?> orderCommand(Map<String, Object?> command) {
    return {
      'schema': 'matchengine.order_command.v2',
      'message_type': 'ORDER',
      'operation': 'ADD',
      ...command,
    };
  }

  Map<String, Object?> marketDataSubscribe(Map<String, Object?> request) {
    return {'schema': 'matchengine.market_data_subscribe.v2', ...request};
  }

  FutureOr<Object?> sendOrder(
    FutureOr<Object?> Function(Map<String, Object?> message) sender,
    Map<String, Object?> command,
  ) {
    return sender(orderCommand(command));
  }

  FutureOr<Object?> subscribeMarketData(
    FutureOr<Object?> Function(Map<String, Object?> message) sender,
    Map<String, Object?> request,
  ) {
    return sender(marketDataSubscribe(request));
  }
}

final class ProviderClient {
  ProviderClient(this.service);

  final ServiceClient service;

  Future<Object?> get(String path) => service.get(path);

  Future<Object?> post(String path, [Object? body]) => service.post(path, body);

  Future<Object?> put(String path, [Object? body]) => service.put(path, body);

  Future<Object?> delete(String path) => service.delete(path);
}

final class WiseClient extends ProviderClient {
  WiseClient(super.service);

  Future<Object?> profiles() => get('/v1/profiles');

  Future<Object?> profile(String profileId) {
    return get('/v1/profiles/${Uri.encodeComponent(profileId)}');
  }

  Future<Object?> createQuote(String profileId, Map<String, Object?> request) {
    return post(
      '/v3/profiles/${Uri.encodeComponent(profileId)}/quotes',
      request,
    );
  }

  Future<Object?> createTransfer(Map<String, Object?> request) {
    return post('/v1/transfers', request);
  }
}

final class AirwallexClient extends ProviderClient {
  AirwallexClient(super.service);

  Future<Object?> accounts() => get('/api/v1/accounts');

  Future<Object?> payments() => get('/api/v1/payments');

  Future<Object?> createPayment(Map<String, Object?> request) {
    return post('/api/v1/payments', request);
  }

  Future<Object?> beneficiaries() => get('/api/v1/beneficiaries');

  Future<Object?> createBeneficiary(Map<String, Object?> request) {
    return post('/api/v1/beneficiaries', request);
  }
}

final class ProvidersClient {
  ProvidersClient({required this.wise, required this.airwallex});

  final WiseClient wise;
  final AirwallexClient airwallex;
}

final class PlannerClient {
  PlannerClient({
    required this.mcp,
    required this.orchestrator,
    required this.analytics,
    required this.tradingControl,
  });

  final MCPClient mcp;
  final OrchestratorClient orchestrator;
  final AnalyticsClient analytics;
  final TradingControlClient tradingControl;

  Future<Object?> conversationPlan(Map<String, Object?> request) {
    return mcp.conversationPlan(request);
  }

  Future<Object?> orchestratedPlan(Map<String, Object?> request) {
    return orchestrator.sendMessage(request);
  }

  Future<Object?> operationalPlan(Map<String, Object?> request) {
    return orchestratedPlan(request);
  }

  Future<Object?> tradingPlan(Map<String, Object?> request) async {
    final existingContext = request['context'];
    final context = <String, Object?>{
      if (existingContext is Map<String, Object?>) ...existingContext,
    };
    if (request['useTrading'] == true || request['use_trading'] == true) {
      try {
        context['trading_robots'] = await tradingControl.robots();
      } catch (_) {
        // Best-effort enrichment keeps planning usable when trading is unavailable.
      }
    }
    if (request['useAnalytics'] == true || request['use_analytics'] == true) {
      try {
        context['analytics_indicators'] = await analytics.indicators();
      } catch (_) {
        // Best-effort enrichment keeps planning usable when analytics is unavailable.
      }
    }
    return conversationPlan({...request, 'context': context});
  }
}

File: python/twofinance_sdk_client/domains.py
from __future__ import annotations

from typing import Any
from urllib.parse import quote

from .service import ServiceClient


def _path(value: str) -> str:
    return quote(value or "", safe="")


class AuthClient(ServiceClient):
    def __init__(
        self,
        base_url: str,
        realm: str = "2finance",
        client_id: str = "2finance-network",
        phone_client_id: str = "2finance-network-phone",
        **kwargs: Any,
    ):
        super().__init__(base_url, **kwargs)
        self.realm = realm
        self.client_id = client_id
        self.phone_client_id = phone_client_id

    def auth_path(self, client_id: str, endpoint: str) -> str:
        return f"/v1/2finance-authenticator/{self.realm}/{client_id}/{endpoint.lstrip('/')}"

    def login(self, username: str, password: str) -> Any:
        return self.post(self.auth_path(self.client_id, "/login"), {"username": username, "password": password})

    def sign_up(self, request: dict[str, Any]) -> Any:
        return self.post(self.auth_path(self.client_id, "/signup"), request)

    def refresh_token(self, refresh_token: str) -> Any:
        return self.post(self.auth_path(self.client_id, "/refresh"), {"refresh_token": refresh_token})

    def phone_login(self, phone_number: str, code: str) -> Any:
        return self.post(
            self.auth_path(self.phone_client_id, "/phone/sms/login"),
            {"phone_number": phone_number, "code": code},
        )

    def jwks(self) -> Any:
        return self.get(self.oidc_path("/protocol/openid-connect/certs"))

    def validate_token(self, token: str) -> Any:
        return self.post(self.oidc_path("/protocol/openid-connect/token/introspect"), {"token": token})

    def oidc_path(self, endpoint: str) -> str:
        return f"/realms/{self.realm}/{endpoint.lstrip('/')}"


class AnalyticsClient(ServiceClient):
    def indicators(self) -> Any:
        return self.get("/analytics/indicators")

    def calculate_technical_analysis(self, request: dict[str, Any]) -> Any:
        return self.post("/analytics/technical-analysis:calculate", request)

    def optimize_portfolio(self, request: dict[str, Any]) -> Any:
        return self.post("/portfolio-manager/optimizer", request)

    def upsert_candles(self, request: dict[str, Any]) -> Any:
        return self.post("/analytics/candles:upsert", request)

    def rankings(self) -> Any:
        return self.get("/portfolio-manager/rankings")

    def balances(self, account_id: str) -> Any:
        return self.get(f"/portfolio-manager/balances/{_path(account_id)}")

    def black_scholes(self, query: str = "") -> Any:
        suffix = f"?{query}" if query else ""
        return self.get(f"/risk-manager/blackscholes{suffix}")

    def staking(self) -> Any:
        return self.get("/staking")


class MCPClient(ServiceClient):
    def __init__(self, base_url: str, **kwargs: Any):
        super().__init__(base_url, **kwargs)
        self._next_id = 1

    def call(self, method: str, params: Any = None) -> Any:
        request_id = self._next_id
        self._next_id += 1
        return self.post("/mcp", {"jsonrpc": "2.0", "id": request_id, "method": method, "params": params})

    def call_tool(self, name: str, arguments: dict[str, Any] | None = None) -> Any:
        return self.call("tools/call", {"name": name, "arguments": arguments or {}})

    def list_tools(self) -> Any:
        return self.call("tools/list")

    def list_prompts(self) -> Any:
        return self.call("prompts/list")

    def list_resources(self) -> Any:
        return self.call("resources/list")

    def read_resource(self, uri: str) -> Any:
        return self.call("resources/read", {"uri": uri})

    def get_prompt(self, name: str, arguments: dict[str, Any] | None = None) -> Any:
        return self.call("prompts/get", {"name": name, "arguments": arguments or {}})

    def conversation_plan(self, arguments: dict[str, Any]) -> Any:
        return self.call_tool("finance_assistant.conversation.plan", arguments)


class OrchestratorClient(ServiceClient):
    def catalog(self) -> Any:
        return self.get("/v1/mcphost/catalog/packages")

    def create_session(self, request: dict[str, Any]) -> Any:
        return self.post("/v1/mcphost/sessions", request)

    def send_message(self, request: dict[str, Any]) -> Any:
        return self.post("/v1/mcphost/messages", request)

    def tools(self) -> Any:
        return self.get("/v1/mcphost/tools")

    def prompts(self) -> Any:
        return self.get("/v1/mcphost/prompts")

    def resources(self) -> Any:
        return self.get("/v1/mcphost/resources")

    def providers(self) -> Any:
        return self.get("/v1/mcphost/providers")

    def approvals(self) -> Any:
        return self.get("/v1/mcphost/approvals")

    def observability(self) -> Any:
        return self.get("/v1/mcphost/observability")

    def delete_session(self, session_id: str) -> Any:
        return self.delete(f"/v1/mcphost/sessions/{_path(session_id)}")


class TradingControlClient(ServiceClient):
    def robots(self) -> Any:
        return self.get("/robots")

    def create_robot(self, request: dict[str, Any]) -> Any:
        return self.post("/robots", request)

    def start_robot(self, robot_id: str) -> Any:
        return self.post(f"/robots/{_path(robot_id)}:start")

    def robot(self, robot_id: str) -> Any:
        return self.get(f"/robots/{_path(robot_id)}")

    def pause_robot(self, robot_id: str) -> Any:
        return self.post(f"/robots/{_path(robot_id)}:pause")

    def resume_robot(self, robot_id: str) -> Any:
        return self.post(f"/robots/{_path(robot_id)}:resume")

    def stop_robot(self, robot_id: str) -> Any:
        return self.post(f"/robots/{_path(robot_id)}:stop")

    def risk_policy(self, robot_id: str) -> Any:
        return self.get(f"/robots/{_path(robot_id)}/risk-policy")

    def set_risk_policy(self, robot_id: str, request: dict[str, Any]) -> Any:
        return self.put(f"/robots/{_path(robot_id)}/risk-policy", request)

    def risk_view(self, robot_id: str) -> Any:
        return self.get(f"/risk-view/{_path(robot_id)}")

    def strategies(self) -> Any:
        return self.get("/strategies")

    def create_strategy(self, request: dict[str, Any]) -> Any:
        return self.post("/strategies", request)

    def directives(self) -> Any:
        return self.get("/directives")

    def create_directive(self, request: dict[str, Any]) -> Any:
        return self.post("/directives", request)

    def audit(self) -> Any:
        return self.get("/audit")

    def activity(self) -> Any:
        return self.get("/activity")

    def mcp_tools(self) -> Any:
        return self.get("/mcp/tools")


class KeyStoreClient(ServiceClient):
    def health(self) -> Any:
        return self.get("/healthz")

    def readiness(self) -> Any:
        return self.get("/readyz")

    def start_keygen(self, request: dict[str, Any]) -> Any:
        return self.post("/keystore/keygen/start", request)

    def start_signing(self, request: dict[str, Any]) -> Any:
        return self.post("/keystore/signing/start", request)

    def keygen_signature(self, request: dict[str, Any]) -> Any:
        return self.post("/keystore/keygen/signature", request)

    def signing_signature(self, request: dict[str, Any]) -> Any:
        return self.post("/keystore/signing/signature", request)

    def start_resharing(self, request: dict[str, Any]) -> Any:
        return self.post("/keystore/resharing/start", request)

    def keys(self, user_public_key: str) -> Any:
        return self.get(f"/keystore/keys/{_path(user_public_key)}")

    def signatures(self, user_public_key: str) -> Any:
        return self.get(f"/keystore/signatures/{_path(user_public_key)}")

    def metrics(self) -> Any:
        return self.get("/keystore/tss/metrics")


class NetworkClient(ServiceClient):
    def virtual_machine(self) -> Any:
        return self.post(
            "/v2/2finance-network/query",
            {"method": "get_blocks", "params": {"page": 1, "limit": 1}},
        )

    def market_candles(self, market: str, query: str = "") -> Any:
        suffix = f"?{query}" if query else ""
        return self.get(f"/v2/2finance-network/markets/{_path(market)}/candles{suffix}")

    def products(self, product_type: str) -> Any:
        return self.get(f"/v2/2finance-network/products/{_path(product_type)}")

    def create_product(self, product_type: str, request: dict[str, Any]) -> Any:
        return self.post(f"/v2/2finance-network/products/{_path(product_type)}", request)

    def bonds(self) -> Any:
        return self.products("bonds")

    def create_bond(self, request: dict[str, Any]) -> Any:
        return self.create_product("bonds", request)

    def loans(self) -> Any:
        return self.products("loans")

    def create_loan(self, request: dict[str, Any]) -> Any:
        return self.create_product("loans", request)

    def swaps(self) -> Any:
        return self.products("swaps")

    def create_swap(self, request: dict[str, Any]) -> Any:
        return self.create_product("swaps", request)

    def staking_products(self) -> Any:
        return self.products("staking")

    def create_staking_product(self, request: dict[str, Any]) -> Any:
        return self.create_product("staking", request)

    def synthetic_assets(self) -> Any:
        return self.products("synthetic-assets")

    def create_synthetic_asset(self, request: dict[str, Any]) -> Any:
        return self.create_product("synthetic-assets", request)

    def liquidity_pools(self) -> Any:
        return self.products("liquidity-pools")

    def create_liquidity_pool(self, request: dict[str, Any]) -> Any:
        return self.create_product("liquidity-pools", request)


class HummingbotClient(ServiceClient):
    def assets(self) -> Any:
        return self.get("/api/v1/assets")

    def symbols(self) -> Any:
        return self.get("/api/v1/symbols")

    def balances(self) -> Any:
        return self.get("/api/v1/balances")

    def connector_config(self, request: dict[str, Any]) -> Any:
        return self.post("/api/v1/connectors/2finance/config", request)


class ProviderClient(ServiceClient):
    pass


class WiseClient(ProviderClient):
    def profiles(self) -> Any:
        return self.get("/v1/profiles")

    def profile(self, profile_id: str) -> Any:
        return self.get(f"/v1/profiles/{_path(profile_id)}")

    def create_quote(self, profile_id: str, request: dict[str, Any]) -> Any:
        return self.post(f"/v3/profiles/{_path(profile_id)}/quotes", request)

    def create_transfer(self, request: dict[str, Any]) -> Any:
        return self.post("/v1/transfers", request)


class AirwallexClient(ProviderClient):
    def accounts(self) -> Any:
        return self.get("/api/v1/accounts")

    def payments(self) -> Any:
        return self.get("/api/v1/payments")

    def create_payment(self, request: dict[str, Any]) -> Any:
        return self.post("/api/v1/payments", request)

    def beneficiaries(self) -> Any:
        return self.get("/api/v1/beneficiaries")

    def create_beneficiary(self, request: dict[str, Any]) -> Any:
        return self.post("/api/v1/beneficiaries", request)


class MatchEngineClient:
    def __init__(self, websocket_url: str):
        self.websocket_url = (websocket_url or "").strip()

    def order_command(self, **command: Any) -> dict[str, Any]:
        payload = {
            "schema": "matchengine.order_command.v2",
            "message_type": "ORDER",
            "operation": "ADD",
        }
        payload.update(command)
        return payload

    def market_data_subscribe(self, **request: Any) -> dict[str, Any]:
        payload = {"schema": "matchengine.market_data_subscribe.v2"}
        payload.update(request)
        return payload

    def send_order(self, sender: Any, **command: Any) -> Any:
        return self._send(sender, self.order_command(**command))

    def subscribe_market_data(self, sender: Any, **request: Any) -> Any:
        return self._send(sender, self.market_data_subscribe(**request))

    @staticmethod
    def _send(sender: Any, message: dict[str, Any]) -> Any:
        if callable(sender):
            return sender(message)
        send = getattr(sender, "send", None)
        if callable(send):
            return send(message)
        raise TypeError("matchengine sender must be callable or expose send(message)")


class PlannerClient:
    def __init__(
        self,
        mcp: MCPClient,
        orchestrator: OrchestratorClient,
        analytics: AnalyticsClient,
        trading_control: TradingControlClient,
    ):
        self.mcp = mcp
        self.orchestrator = orchestrator
        self.analytics = analytics
        self.trading_control = trading_control

    def conversation_plan(self, request: dict[str, Any]) -> Any:
        return self.mcp.conversation_plan(request)

    def orchestrated_plan(self, request: dict[str, Any]) -> Any:
        return self.orchestrator.send_message(request)

    def operational_plan(self, request: dict[str, Any]) -> Any:
        return self.orchestrated_plan(request)

    def trading_plan(self, request: dict[str, Any]) -> Any:
        context = dict(request.get("context") or {})
        if request.get("use_trading") is True or request.get("useTrading") is True:
            try:
                context["trading_robots"] = self.trading_control.robots()
            except Exception:
                # Best-effort enrichment keeps planning usable when trading is unavailable.
                pass
        if request.get("use_analytics") is True or request.get("useAnalytics") is True:
            try:
                context["analytics_indicators"] = self.analytics.indicators()
            except Exception:
                # Best-effort enrichment keeps planning usable when analytics is unavailable.
                pass
        return self.conversation_plan({**request, "context": context})
